File: Sources/XMTP/Messages/MessageV2.swift
//
//  MessageV2.swift
//
//
//  Created by Pat Nakajima on 12/5/22.
//

import CryptoKit
import Foundation
import XMTPRust

typealias MessageV2 = Xmtp_MessageContents_MessageV2

enum MessageV2Error: Error {
	case invalidSignature, decodeError(String)
}

extension MessageV2 {
	init(headerBytes: Data, ciphertext: CipherText) {
		self.init()
		self.headerBytes = headerBytes
		self.ciphertext = ciphertext
	}

	static func decrypt(_ id: String, _ topic: String, _ message: MessageV2, keyMaterial: Data, client: Client) throws -> DecryptedMessage {
		let decrypted = try Crypto.decrypt(keyMaterial, message.ciphertext, additionalData: message.headerBytes)
		let signed = try SignedContent(serializedData: decrypted)

		guard signed.sender.hasPreKey, signed.sender.hasIdentityKey else {
			throw MessageV2Error.decodeError("missing sender pre-key or identity key")
		}

		let senderPreKey = try PublicKey(signed.sender.preKey)
		let senderIdentityKey = try PublicKey(signed.sender.identityKey)

		// This is a bit confusing since we're passing keyBytes as the digest instead of a SHA256 hash.
		// That's because our underlying crypto library always SHA256's whatever data is sent to it for this.
		if !(try senderPreKey.signature.verify(signedBy: senderIdentityKey, digest: signed.sender.preKey.keyBytes)) {
			throw MessageV2Error.decodeError("pre-key not signed by identity key")
		}

		// Verify content signature
		let key = try PublicKey.with { key in
			key.secp256K1Uncompressed.bytes = try KeyUtilx.recoverPublicKeySHA256(from: signed.signature.rawData, message: Data(message.headerBytes + signed.payload))
		}

		if key.walletAddress != (try PublicKey(signed.sender.preKey).walletAddress) {
			throw MessageV2Error.invalidSignature
		}

		let encodedMessage = try EncodedContent(serializedData: signed.payload)
		let header = try MessageHeaderV2(serializedData: message.headerBytes)

		return DecryptedMessage(
			id: id,
			encodedContent: encodedMessage,
			senderAddress: try signed.sender.walletAddress,
			sentAt: Date(timeIntervalSince1970: Double(header.createdNs / 1_000_000) / 1000),
			topic: topic
		)
	}

	static func decode(_ id: String, _ topic: String, _ message: MessageV2, keyMaterial: Data, client: Client) throws -> DecodedMessage {
		do {
			let decryptedMessage = try decrypt(id, topic, message, keyMaterial: keyMaterial, client: client)

			return DecodedMessage(
				id: id,
				client: client,
				topic: decryptedMessage.topic,
				encodedContent: decryptedMessage.encodedContent,
				senderAddress: decryptedMessage.senderAddress,
				sent: decryptedMessage.sentAt
			)
		} catch {
			print("ERROR DECODING: \(error)")
			throw error
		}
	}

	static func encode(client: Client, content encodedContent: EncodedContent, topic: String, keyMaterial: Data) async throws -> MessageV2 {
		let payload = try encodedContent.serializedData()

		let date = Date()
		let header = MessageHeaderV2(topic: topic, created: date)
		let headerBytes = try header.serializedData()

		let digest = SHA256.hash(data: headerBytes + payload)
		let preKey = client.keys.preKeys[0]
		let signature = try await preKey.sign(Data(digest))

		let bundle = client.privateKeyBundleV1.toV2().getPublicKeyBundle()

		let signedContent = SignedContent(payload: payload, sender: bundle, signature: signature)
		let signedBytes = try signedContent.serializedData()

		let ciphertext = try Crypto.encrypt(keyMaterial, signedBytes, additionalData: headerBytes)

		return MessageV2(
			headerBytes: headerBytes,
			ciphertext: ciphertext
		)
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/MessageV2.kt
package org.xmtp.android.library.messages

import com.google.protobuf.kotlin.toByteString
import com.google.protobuf.kotlin.toByteStringUtf8
import org.web3j.crypto.ECDSASignature
import org.web3j.crypto.Hash
import org.web3j.crypto.Sign
import org.xmtp.android.library.CipherText
import org.xmtp.android.library.Client
import org.xmtp.android.library.Crypto
import org.xmtp.android.library.DecodedMessage
import org.xmtp.android.library.KeyUtil
import org.xmtp.android.library.XMTPException
import org.xmtp.android.library.codecs.ContentCodec
import org.xmtp.android.library.codecs.EncodedContent
import java.math.BigInteger
import java.util.Date

typealias MessageV2 = org.xmtp.proto.message.contents.MessageOuterClass.MessageV2

class MessageV2Builder(val senderHmac: ByteArray? = null, val shouldPush: Boolean = false) {
    lateinit var messageV2: MessageV2

    companion object {
        fun buildFromCipherText(
            headerBytes: ByteArray,
            ciphertext: CipherText?,
            senderHmac: ByteArray,
            shouldPush: Boolean,
        ): MessageV2Builder {
            val messageBuilder = MessageV2Builder(senderHmac = senderHmac, shouldPush = shouldPush)
            messageBuilder.messageV2 = MessageV2.newBuilder().also {
                it.headerBytes = headerBytes.toByteString()
                it.ciphertext = ciphertext
                it.shouldPush = shouldPush
                it.senderHmac = senderHmac.toByteString()
            }.build()
            return messageBuilder
        }

        fun buildDecode(
            id: String,
            topic: String,
            message: MessageV2,
            keyMaterial: ByteArray,
            client: Client,
        ): DecodedMessage {
            try {
                val decryptedMessage = buildDecrypt(id, topic, message, keyMaterial, client)

                return DecodedMessage(
                    id = id,
                    client = client,
                    topic = decryptedMessage.topic,
                    encodedContent = decryptedMessage.encodedContent,
                    senderAddress = decryptedMessage.senderAddress,
                    sent = decryptedMessage.sentAt,
                )
            } catch (e: Exception) {
                throw XMTPException("Error decoding message", e)
            }
        }

        fun buildDecrypt(
            id: String,
            topic: String,
            message: MessageV2,
            keyMaterial: ByteArray,
            client: Client,
        ): DecryptedMessage {
            val decrypted =
                Crypto.decrypt(keyMaterial, message.ciphertext, message.headerBytes.toByteArray())
            val signed = SignedContent.parseFrom(decrypted)

            if (!signed.sender.hasPreKey() || !signed.sender.hasIdentityKey()) {
                throw XMTPException("missing sender pre-key or identity key")
            }

            val senderPreKey = PublicKeyBuilder.buildFromSignedPublicKey(signed.sender.preKey)
            val senderIdentityKey =
                PublicKeyBuilder.buildFromSignedPublicKey(signed.sender.identityKey)

            if (!senderPreKey.signature.verify(
                    senderIdentityKey,
                    signed.sender.preKey.keyBytes.toByteArray(),
                )
            ) {
                throw XMTPException("pre-key not signed by identity key")
            }

            // Verify content signature
            val digest =
                Hash.sha256(message.headerBytes.toByteArray() + signed.payload.toByteArray())

            val signatureData =
                KeyUtil.getSignatureData(signed.signature.rawData.toByteString().toByteArray())
            val publicKey = Sign.recoverFromSignature(
                BigInteger(1, signatureData.v).toInt(),
                ECDSASignature(BigInteger(1, signatureData.r), BigInteger(1, signatureData.s)),
                digest,
            )

            val key = PublicKey.newBuilder().also {
                it.secp256K1Uncompressed = it.secp256K1Uncompressed.toBuilder().also { keyBuilder ->
                    keyBuilder.bytes =
                        KeyUtil.addUncompressedByte(publicKey.toByteArray()).toByteString()
                }.build()
            }.build()

            if (key.walletAddress != (PublicKeyBuilder.buildFromSignedPublicKey(signed.sender.preKey).walletAddress)) {
                throw XMTPException("Invalid signature")
            }

            val encodedMessage = EncodedContent.parseFrom(signed.payload)
            val header = MessageHeaderV2.parseFrom(message.headerBytes)
            if (header.topic != topic) {
                throw XMTPException("Topic mismatch")
            }

            return DecryptedMessage(
                id = id,
                encodedContent = encodedMessage,
                senderAddress = signed.sender.walletAddress,
                sentAt = Date(header.createdNs / 1_000_000),
                topic = topic,
            )
        }

        private fun <Codec : ContentCodec<T>, T> shouldPush(codec: Codec, content: T?): Boolean {
            if (content != null) {
                return codec.shouldPush(content = content)
            } else {
                throw XMTPException("Codec invalid content")
            }
        }

        fun <Codec : ContentCodec<T>, T> buildEncode(
            client: Client,
            encodedContent: EncodedContent,
            topic: String,
            keyMaterial: ByteArray,
            codec: Codec,
        ): MessageV2Builder {
            val payload = encodedContent.toByteArray()
            val date = Date()
            val header = MessageHeaderV2Builder.buildFromTopic(topic, date)
            val headerBytes = header.toByteArray()
            val digest = Hash.sha256(headerBytes + payload)
            val preKey = client.keys.preKeysList?.get(0)
            val signature = preKey?.sign(digest)
            val bundle = client.privateKeyBundleV1.toV2().getPublicKeyBundle()
            val signedContent = SignedContentBuilder.builderFromPayload(payload, bundle, signature)
            val signedBytes = signedContent.toByteArray()
            val ciphertext = Crypto.encrypt(keyMaterial, signedBytes, additionalData = headerBytes)

            val thirtyDayPeriodsSinceEpoch =
                (Date().time / 1000 / 60 / 60 / 24 / 30).toInt()
            val info = "$thirtyDayPeriodsSinceEpoch-${client.address}"
            val infoEncoded = info.toByteStringUtf8().toByteArray()
            val senderHmacGenerated =
                Crypto.calculateMac(
                    Crypto.deriveKey(keyMaterial, ByteArray(0), infoEncoded),
                    headerBytes
                )

            return buildFromCipherText(
                headerBytes,
                ciphertext,
                senderHmacGenerated,
                shouldPush(codec = codec, content = codec.decode(encodedContent)),
            )
        }
    }
}
